crates/compiler/src/lib.rs

```diff
@@ -31,35 +30,35 @@ pub mod solc;
 
 /// A common interface for all supported Solidity compilers.
 pub trait SolidityCompiler {
-    /// Extra options specific to the compiler.
-    type Options: Default + PartialEq + Eq + Hash;
-
-    /// The low-level compiler interface.
-    fn build(
-        &self,
-        input: CompilerInput,
-        additional_options: Self::Options,
-    ) -> impl Future<Output = anyhow::Result<CompilerOutput>>;
+    /// Instantiates a new compiler object.
+    ///
+    /// Based on the given [`Arguments`] and [`VersionOrRequirement`] this function instantiates a
+    /// new compiler object. Certain implementations of this trait might choose to cache cache the
+    /// compiler objects and return the same ones over and over again.
+    fn new(
+        config: &Arguments,
+        version: impl Into<Option<VersionOrRequirement>>,
+    ) -> impl Future<Output = Result<Arc<Self>>>;
```

Comment:
Nit: Normally I'd consider hiding the `Arc` inside the abstraction, so that each concrete compiler can decode on how to best let itself be cloned. Either way works though